pkg/config/xds/filter_types.go

```diff
@@ -12,32 +15,47 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// nolint: lll
-//
-//go:generate sh -c "echo '//go:build !agent' > filter_types.gen.go"
-//go:generate sh -c "echo '// +build !agent\n' >> filter_types.gen.go"
-//go:generate sh -c "echo '// Copyright Istio Authors' >> filter_types.gen.go"
-//go:generate sh -c "echo '//' >> filter_types.gen.go"
-//go:generate sh -c "echo '// Licensed under the Apache License, Version 2.0 (the \"License\");' >> filter_types.gen.go"
-//go:generate sh -c "echo '// you may not use this file except in compliance with the License.' >> filter_types.gen.go"
-//go:generate sh -c "echo '// You may obtain a copy of the License at' >> filter_types.gen.go"
-//go:generate sh -c "echo '//' >> filter_types.gen.go"
-//go:generate sh -c "echo '//     http://www.apache.org/licenses/LICENSE-2.0' >> filter_types.gen.go"
-//go:generate sh -c "echo '//' >> filter_types.gen.go"
-//go:generate sh -c "echo '// Unless required by applicable law or agreed to in writing, software' >> filter_types.gen.go"
-//go:generate sh -c "echo '// distributed under the License is distributed on an \"AS IS\" BASIS,' >> filter_types.gen.go"
-//go:generate sh -c "echo '// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.' >> filter_types.gen.go"
-//go:generate sh -c "echo '// See the License for the specific language governing permissions and' >> filter_types.gen.go"
-//go:generate sh -c "echo '// limitations under the License.\n' >> filter_types.gen.go"
-//go:generate sh -c "echo '// Code generated by filter_types.go. DO NOT EDIT.\n' >> filter_types.gen.go"
-//go:generate sh -c "echo 'package xds\n\nimport (' >> filter_types.gen.go"
-//go:generate sh -c "go list github.com/envoyproxy/go-control-plane/... | grep 'v3' | grep -v /pkg/ | xargs -I{} echo '\t_ \"{}\"' >> filter_types.gen.go"
-//go:generate sh -c "echo '\n\t// Istio-specific Envoy filters' >> filter_types.gen.go"
-//go:generate sh -c "go list istio.io/api/envoy/config/filter/... | grep 'v[0-9]' | xargs -I{} echo '\t_ \"{}\"' >> filter_types.gen.go"
-//go:generate sh -c "echo ')' >> filter_types.gen.go"
 package xds
 
-// Import all Envoy filter types so they are registered and deserialization does not fail
-// when using them in the "typed_config" attributes.
-// The filter types are autogenerated by looking at all packages in go-control-plane
-// As a result, this will need to be re-run when updating go-control-plane if new packages are added.
+// Import all *common* Envoy filter types so they are registered.
+// When registered, we can do full validation and more efficient conversion.
+// Otherwise, we fallback to TypedStruct.
```

Comment:
What's the performance influence it brings other than the binary size?

Reply:
There are a few parts here.

1. Fallback to TypedStruct, in general. Using TypedStruct instead of Any adds about 10% to the proto size in typical cases I measured, and has a small cost in the control plane to do the conversion. At runtime in Envoy (filter execution), the performance is identical -- only xDS is impacted.

The benefit of this is not performance, but compatibility. Today, a user can only use the protos that happen to be compiled into Istiod. For example, if you have a custom envoy with some filters we don't know about, or we just missed some filters, etc -- this PR makes them work; before they would be rejected by Istiod.

2. Moving from Istiod importing _all_ envoy protos to just the most common ones. This has a few benefits:
* Primary benefit is not having a constantly bloating and changing filter_types.gen.go. We import _every_ proto in Envoy, which is a list that is constantly growing and will never stop growing. Currently there are ~300 proto packages we import! Some of these are extremely niche like `github.com/envoyproxy/go-control-plane/contrib/envoy/extensions/compression/qatzip/compressor/v3alpha` -- an alpha compression plugin for a specific type of hardware. While we have some nice automation that makes this ok to keep up with, it still doesn't feel great to so tightly couple Istio to these
* 10Mb smaller binary, 2x faster binary startup (mostly useful for tests), 10% faster compile time. Not a huge motivator, but nice to have. Note the 10mb smaller binary results in a similar ~10mb reduction in RSS usage during runtime

Reply:
I am not an expert on this, but would this increase the risk of invalid envoyfilter patching